cmd/k8srules/controller.go

```diff
@@ -0,0 +1,273 @@
+		cacheResyncPeriod,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    workqueue.EnqueueingAddFunc(controller.addRule),
+			UpdateFunc: workqueue.EnqueueingUpdateFunc(controller.updateRule),
+		},
+	)
+
+	controller.workqueue.Start()
+	go controller.rulesController.Run(ctx.Done())
+
+	return controller, nil
+}
+
+// Stop synchronizing the rules.
+func (c *controller) Stop() error {
+	c.workqueue.Stop()
+
+	return nil
+}
+
+// initialize third party resource if it does not exist
+func initThirdPartyResource() error {
+	client, err := kubepkg.NewClient(kubepkg.Config{})
+	if err != nil {
+		return err
+	}
+
+	resName := fmt.Sprintf("%s.%s", kuberules.ResourceName, kuberules.ResourceGroupName)
+	_, err = client.Extensions().ThirdPartyResources().Get(resName)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			tpr := &v1beta1.ThirdPartyResource{
+				ObjectMeta: v1.ObjectMeta{
+					Name: resName,
+				},
+				Versions: []v1beta1.APIVersion{
+					{Name: kuberules.ResourceVersion},
+				},
+				Description: kuberules.ResourceDescription,
+			}
+
+			result, err := client.Extensions().ThirdPartyResources().Create(tpr)
+			if err != nil {
+				return err
+			}
+			log.Printf("The ThridPartYresource %#v created\n", result)
+		} else {
+			return err
+		}
+	} else {
+		log.Printf("The ThirdPartyResource %s already exists\n", resName)
+	}
+	return nil
+}
+
+func createTPRClient() (*rest.RESTClient, error) {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		log.Printf("Failed to create the in-cluster config. %s\n", err)
+		return nil, err
+	}
+
+	groupversion := unversioned.GroupVersion{
+		Group:   kuberules.ResourceGroupName,
+		Version: kuberules.ResourceVersion,
+	}
+
+	config.GroupVersion = &groupversion
+	config.APIPath = "/apis"
+	config.ContentType = runtime.ContentTypeJSON
+	config.NegotiatedSerializer = serializer.DirectCodecFactory{CodecFactory: api.Codecs}
+
+	schemeBuilder := runtime.NewSchemeBuilder(
+		func(scheme *runtime.Scheme) error {
+			scheme.AddKnownTypes(
+				groupversion,
+				&kuberules.Rule{},
+				&kuberules.RuleList{},
+				&api.ListOptions{},
+				&api.DeleteOptions{},
+			)
+			return nil
+		})
+	schemeBuilder.AddToScheme(api.Scheme)
+
+	client, err := rest.RESTClientFor(config)
+	if err != nil {
+		log.Printf("Failed to create the in-cluster client. %s\n", err)
+		return nil, err
+	}
+
+	return client, nil
+}
+
+func (c *controller) validateRule(rule *kuberules.Rule) error {
+	var updRule kuberules.Rule
+	var result kuberules.Rule
+
+	log.Printf("Rule %s is being validated\n", rule.Metadata.Name)
+
+	if rule.Spec.ID == "" {
+		updRule.Spec.ID = rule.Metadata.Name
+	}
+
+	err := c.validator.Validate(rule.Spec)
+	if err != nil {
+		log.Printf("Rule %#v is not valid. %s\n", rule, err)
+		updRule.Status.State = kuberules.RuleStateInvalid
+		updRule.Status.Message = err.Error()
+	} else {
+		log.Printf("Rule %s is valid\n", rule.Metadata.Name)
+		updRule.Status.State = kuberules.RuleStateValid
+	}
```

Comment:
By delta, I mean will it be just that line in JSON or will the sidecar receive the entire Rules Array that is stuffed into the TPR? Any chance that this can be compressed ?